### PR TITLE
Port the rules evaluator from internal repo

### PR DESCRIPTION
**Pull request by dagnir**
This commit ports the internal rule repository from the internal repository.

Note: The files will only live in `sdk-core` during development. Similarly, the new `json-util` dependency introduced is development only.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/sdk-core/src/test/java/software/amazon/awssdk/core/rules/IntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    @ParameterizedTest
+    @MethodSource("checkableTestCases")
+    void executeTestSuite(TestDiscovery.RulesTestcase testcase) {
+        testcase.testcase().execute(testcase.ruleset());
+    }
+
+    private Stream<ValidationTestCase> validTestcases() {
+        TestDiscovery discovery = new TestDiscovery();
```
Comment (cenedhryn) [inline core/sdk-core/src/test/java/software/amazon/awssdk/core/rules/IntegrationTest.java:53]: It's confusing to have two instances of TestDiscovery, is it possible to refactor?

**Review by cenedhryn — COMMENTED** [review]
General comment: For code that has a long-term prospect in the SDK I'd wish for more javadocs at some point in time

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/rules/RuleError.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    public static Builder builder() {
+        return new BuilderImpl();
+    }
+
+    public interface Builder extends SdkException.Builder {
+        @Override
+        RuleError build();
+    }
+
+    public static <T> T ctx(String message, Supplier<T> f) {
```
[the remark this answers is not in the document]
Comment (dagnir) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/rules/RuleError.java:38]: Yeah I actually don't think we need this, I'll remove it in a future PR.

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/rules/Value.java [hunk 3]
```diff
@@ -15,52 +15,418 @@
...
+                        b.property(k, Value.fromNode(v));
+                    });
+            }
+
+            JsonNode headersNode = objNode.get(HEADERS);
+            if (headersNode != null) {
+                headersNode.asObject()
+                    .forEach((k, v) -> v.asArray().forEach(e -> b.addHeader(k, e.asString())));
+            }
+            // ObjectNode on = node.expectObjectNode("endpoints are object nodes");
```
[the remark this answers is not in the document]
Comment (dagnir) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/rules/Value.java:371]: Yes, this was the old implementation using Smithy classes.

File: core/sdk-core/src/test/java/software/amazon/awssdk/core/rules/testutil/TestDiscovery.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import software.amazon.awssdk.core.rules.EndpointRuleset;
+import software.amazon.awssdk.core.rules.EndpointTest;
+import software.amazon.awssdk.core.rules.EndpointTestSuite;
+import software.amazon.awssdk.protocols.jsoncore.JsonNode;
+
+public class TestDiscovery {
```
[the remark this answers is not in the document]
Comment (dagnir) [reply core/sdk-core/src/test/java/software/amazon/awssdk/core/rules/testutil/TestDiscovery.java:20]: Russell wrote this originally! I added some refactoring during the port :).

File: core/sdk-core/src/test/java/software/amazon/awssdk/core/rules/IntegrationTest.java (hunk 1 above)
Comment (dagnir) [reply core/sdk-core/src/test/java/software/amazon/awssdk/core/rules/IntegrationTest.java:53]: Oh good catch. Removed the local var.

**Review by cenedhryn — APPROVED** [review]

**Pull request merged**